datafusion/functions/src/math/round.rs

```diff
@@ -114,7 +116,11 @@ pub fn round(args: &[ArrayRef]) -> Result<ArrayRef> {
     match args[0].data_type() {
         DataType::Float64 => match decimal_places {
             ColumnarValue::Scalar(ScalarValue::Int64(Some(decimal_places))) => {
-                let decimal_places = decimal_places.try_into().unwrap();
+                let decimal_places: i32 = decimal_places.try_into().map_err(|e| {
```

Comment:
I read this change more carefully -- it seems like what you are tring to do is to convert an argument that is i64 into i32

However, this is basically undoing the work higher up in the planner that ensures round is called with an `Int64` as a second argument
https://github.com/apache/datafusion/blob/459afbb3a180d31e7cdefffb46f033069aa47408/datafusion/functions/src/math/round.rs#L48-L49

So if the goal is to get the second argument as `Int32` I think you could just change the desired signature

Reply:
thx for getting back to this! Well yeah, changing the signature is the longer path I initially didn't want to take -- because, afaiu, the integer literals are currently always Int64, so no matter how large the input is, planner always throws an error if just the signature is changed.

I looked a bit into what could be done, and there are some places where type coercion is attempted, like [here](https://github.com/apache/datafusion/blob/main/datafusion/expr/src/type_coercion/functions.rs#L408), but thats on data types only so does upcasts, whereas I'd need a downcast-if-fits. The place where values are "lost" and only types are propagated further during planning is here https://github.com/apache/datafusion/blob/main/datafusion/expr/src/expr_schema.rs#L142

I could extend the latter call to carry both values and types, and then the `coerced_from` to consider downcasting as well if the value fits, if you think this is the reasonable path. Sounds like "still-smallish" PR to me, though not 100% sure it would work as I imagine it to

Alternatively, I could investigate changing the literal parsing itself, to not have always Int64, but instead pick the smallest type possible (this should be also ok, because functions with signatures like Int64 would still accept Int8 etc, as upcasts are already in the code). I haven't yet checked in the code what this would entail

Both option sound legit to me from UX PoV, no preference on my side 🤷

Reply:
thx for finishing this @alamb ! The error you encountered after changing the signature is exactly what I meant, sorry for my convoluted description O:) I see the fix you merged keeps the signature as Int64 -- do you think I should peek more into the alternatives I described above to make it overall more cleaner, ie, make the query planner aware of the actual signature? (Though I'd happily look into other issues if this one isn't really worth it)

Reply:
> I see the fix you merged keeps the signature as Int64 -- do you think I should peek more into the alternatives I described above to make it overall more cleaner, ie, make the query planner aware of the actual signature? 

I personally think it is fine the way it is now -- I don't think there is any reason to make the implementation more complicated unless we have some evidence we need to optimize further (e.g. a benchmark or something)

I expect the common case is that `decimal_places` is a scalar so this will be just fine